File: lib/subsets.rb
def find_subsets(nums)
  nums.combination(4)
  # [1, 2, 3, 4]
  subsets = []
  0.upto(nums.length) { |i| subsets << find_subsets_by_length(nums, i) }
end

def find_subsets_by_length(nums, length, target = [], index = 0)
  #base case
  if index == length-1
    require 'pry'; binding.pry
  end

  #recursive case
  for i in index...nums.size
    target << index
    find_subsets_by_length(nums, length, target, index + 1)
  end
end

def generate_subsets(nums)
  if !arr
    return [[]]
  end

  subset = generate_subsets(nums[1..])
  return subset + [[nums[0]] + s 
end

File: lib/subsets.py
def find_subsets(array):
  output = [[]]
  for i, number in enumerate(array):
    if [number] not in output:
      output.append([number])
    for second_number in array[(i+1):]:
      if [number, second_number] not in output:
        output.append([number, second_number])
    for index in range(i+1, len(array)):
      x = array[i:index+1]
      y = array[:i] + array[index:len(array)]
      if x not in output:
        output.append(x)
      if y not in output:
        output.append(y)
  return print(sorted(output, key=lambda x: (len(x), x)))

print(find_subsets([1, 2, 3, 4, 5]))

def generate_subsets(arr):
  if not arr:
    return [[]] # return an empty subset if the array is empty

  subset = generate_subsets(arr[1:]) # generate all subsets of the remaining elements
  return subset + [[arr[0]] + s for s in subset] # combine the subsets with the first element and without the first element


def sort_subsets(subsets):
  return print(sorted(subsets, key=lambda x: (len(x), x)))